// rocket-punch/package-json/computePackageJson.js
export async function computePackageJson({ packageInfo, dependencies, sharedConfig = {}, }) {
    const shared = { ...sharedConfig };
    const keys = Object.keys(shared);
    keys.forEach((key) => {
        const value = shared[key];
        if (typeof value === 'string') {
            //@ts-ignore
            shared[key] = value
                .replace(/({name})/g, packageInfo.name) // {name}
                .replace(/({version})/g, packageInfo.version); // {version}
        }
    });
    const computedConfig = {
        ...shared,
        //main: 'index.js',
        typings: 'index.d.ts',
        ...packageInfo.packageJson,
        name: packageInfo.name,
        version: packageInfo.version,
        dependencies: dependencies,
    };
    const commonjsDirectory = packageInfo.exports.main === 'commonjs' ? '.' : './_commonjs';
    computedConfig.main = `${commonjsDirectory}/index.js`;
    if (packageInfo.exports.module) {
        const moduleDirectory = packageInfo.exports.main === 'module' ? '.' : './_module';
        computedConfig.module = `${moduleDirectory}/index.js`;
    }
    return computedConfig;
}
