// rocket-punch/rule/getCompilerOptions.js
import { parseTSConfig, readTSConfig } from '@ssen/read-tsconfig';
import fs from 'fs-extra';
import path from 'path';
import ts from 'typescript';
export function getCompilerOptions({ searchPath, configName, packageInfo, buildType, declaration, }) {
    const { options: tsconfig } = fs.existsSync(path.join(searchPath, configName))
        ? readTSConfig(searchPath, configName)
        : { options: {} };
    const { options: info } = parseTSConfig(searchPath, {
        compilerOptions: packageInfo.compilerOptions,
    });
    const defaultValues = {
        downlevelIteration: true,
        allowSyntheticDefaultImports: true,
        esModuleInterop: true,
        strict: true,
        resolveJsonModule: true,
        allowJs: true,
        jsx: ts.JsxEmit.React,
        target: ts.ScriptTarget.ES2018,
    };
    const computed = Object.keys(defaultValues).reduce((result, name) => {
        var _a, _b;
        result[name] = (_b = (_a = info[name]) !== null && _a !== void 0 ? _a : tsconfig[name]) !== null && _b !== void 0 ? _b : defaultValues[name];
        return result;
    }, {});
    return {
        ...computed,
        module: buildType === 'module' ? ts.ModuleKind.ES2020 : ts.ModuleKind.CommonJS,
        moduleResolution: ts.ModuleResolutionKind.NodeJs,
        skipLibCheck: true,
        inlineSources: true,
        inlineSourceMap: true,
        declaration,
    };
}
