// rocket-punch/_commonjs/package-json/getExports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExports = void 0;
const promised_1 = require("@ssen/promised");
async function getExports(exports, outDir) {
    const result = await promised_1.glob(`**/*`, {
        cwd: outDir,
        nodir: true,
        ignore: ['**/*.d.ts', '**/*.js.map', 'public/*', 'bin/*'],
    });
    const files = new Set(result);
    const moduleDirectory = exports.main === 'module' ? '.' : './_module';
    const commonjsDirectory = exports.main === 'module' ? './_commonjs' : '.';
    const secondary = exports.main === 'module' ? 'commonjs' : 'module';
    const entry = result.filter((file) => !file.startsWith('_commonjs') &&
        !file.startsWith('_module') &&
        file !== 'index.js');
    const entryExports = {};
    for (const file of entry) {
        const key = `./${file.replace(/.js$/, '')}`;
        if (!files.has(`_${secondary}/${file}`)) {
            entryExports[key] = `./${file}`;
            continue;
        }
        entryExports[key] = {
            import: `${moduleDirectory}/${file}`,
            require: `${commonjsDirectory}/${file}`,
            node: `${commonjsDirectory}/${file}`,
            default: `${moduleDirectory}/${file}`,
        };
        entryExports[key + '.js'] = {
            import: `${moduleDirectory}/${file}`,
            require: `${commonjsDirectory}/${file}`,
            node: `${commonjsDirectory}/${file}`,
            default: `${moduleDirectory}/${file}`,
        };
        if (/\/index$/.test(key)) {
            entryExports[key.replace(/\/index$/, '')] = {
                import: `${moduleDirectory}/${file}`,
                require: `${commonjsDirectory}/${file}`,
                node: `${commonjsDirectory}/${file}`,
                default: `${moduleDirectory}/${file}`,
            };
        }
    }
    return {
        '.': {
            import: `${moduleDirectory}/index.js`,
            require: `${commonjsDirectory}/index.js`,
            node: `${commonjsDirectory}/index.js`,
            default: `${moduleDirectory}/index.js`,
        },
        './public/*': './public/*',
        ...entryExports,
    };
}
exports.getExports = getExports;
