// @ssen/publish-packages/getPublishOptions.js
import { flatPackageName } from '@ssen/flat-package-name';
import fs from 'fs-extra';
import getPackageJson from 'package-json';
import path from 'path';
const getNpmRemotePackageJson = ({ name, ...options }) => {
    return getPackageJson(name, options)
        .then((value) => value && typeof value.version === 'string'
        ? value
        : undefined)
        .catch(() => undefined);
};
export async function getPublishOptions({ packages, outDir, tag: forceTag, registry: forceRegistry, getRemotePackageJson = getNpmRemotePackageJson, }) {
    if (!fs.existsSync(outDir) || !fs.statSync(outDir).isDirectory()) {
        throw new Error(`"${outDir}" directory is not exists`);
    }
    const tags = new Map();
    for (const [name, { tag }] of packages) {
        tags.set(name, tag !== null && tag !== void 0 ? tag : 'latest');
    }
    const currentPackageJsons = Array.from(packages.values())
        // PackageInfo => /path/to/dist/{name}/package.json
        .map(({ name: packageName }) => path.join(outDir, flatPackageName(packageName), 'package.json'))
        // /path/to/dist/{name}/package.json => boolean
        .filter((packageJsonFile) => fs.existsSync(packageJsonFile))
        // /path/to/dist/{name}/package.json => PackageJson
        .map((packageJsonFile) => fs.readJsonSync(packageJsonFile))
        // PackageJson => boolean
        .filter(({ name }) => typeof name === 'string');
    const remotePackageJsons = await Promise.all(currentPackageJsons.map(({ name }) => {
        if (!name)
            throw new Error(``);
        return getRemotePackageJson({
            name,
            version: forceTag || tags.get(name),
            registryUrl: forceRegistry,
            fullMetadata: true,
        });
    }));
    return currentPackageJsons.reduce((map, current, i) => {
        if (!current || !current.name) {
            throw new Error(``);
        }
        map.set(current.name, {
            name: current.name,
            tag: tags.get(current.name),
            current,
            remote: remotePackageJsons[i],
        });
        return map;
    }, new Map());
}
