// @ssen/publish-packages/selectPublishOptions.js
import prompts from 'prompts';
import semver from 'semver';
export function getVersions({ current, remote }) {
    const currentVersion = current.version;
    const remoteVersion = remote && typeof remote.version === 'string' ? remote.version : undefined;
    return { currentVersion, remoteVersion };
}
export async function selectPublishOptions({ publishOptions, skipSelection, }) {
    const availablePublishOptions = Array.from(publishOptions.values())
        .map((publishOption) => ({
        ...publishOption,
        ...getVersions(publishOption),
    }))
        .filter(({ currentVersion, remoteVersion }) => !remoteVersion || semver.gt(currentVersion, remoteVersion));
    if (availablePublishOptions.length === 0) {
        return [];
    }
    else if (skipSelection) {
        return availablePublishOptions;
    }
    else {
        const answer = await prompts({
            type: 'multiselect',
            name: 'publishOptions',
            message: 'Select packages to publish',
            choices: availablePublishOptions.map(({ name, tag, currentVersion, remoteVersion }) => {
                return {
                    title: remoteVersion
                        ? `${name}@${tag} (${remoteVersion} → ${currentVersion})`
                        : `${name}@${tag} (→ ${currentVersion})`,
                    value: name,
                    disabled: remoteVersion && semver.lte(currentVersion, remoteVersion),
                };
            }),
        });
        const filter = new Set(answer.publishOptions);
        return Array.from(availablePublishOptions.values()).filter((publishOption) => filter.has(publishOption.name));
    }
}
