// @ssen/extended-compiler-host/_commonjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createExtendedCompilerHost = exports.targetExtensions = void 0;
const transform_1 = require("@ssen/transform");
const typescript_1 = __importDefault(require("typescript"));
const transformerMap = {
    html: transform_1.plainTextTransformer,
    ejs: transform_1.plainTextTransformer,
    txt: transform_1.plainTextTransformer,
    md: transform_1.plainTextTransformer,
    yml: transform_1.yamlTransformer,
    yaml: transform_1.yamlTransformer,
    jpg: transform_1.imageTransformer,
    jpeg: transform_1.imageTransformer,
    gif: transform_1.imageTransformer,
    png: transform_1.imageTransformer,
    webp: transform_1.imageTransformer,
    svg: transform_1.svgTransformer,
};
exports.targetExtensions = Object.keys(transformerMap);
function findConfig(fileName) {
    for (const ext of exports.targetExtensions) {
        if (new RegExp(`\\.${ext}\\.tsx$`).test(fileName)) {
            return transformerMap[ext];
        }
    }
    return undefined;
}
function createExtendedCompilerHost(options, setParentNodes, compilerHost = typescript_1.default.createCompilerHost(options, setParentNodes)) {
    // When user import like `import svg from './some.svg'
    // CompilerHost searches by fileExists()
    // /absoulte-path/some.svg.ts -> /absoulte-path/some.svg.tsx -> /absoulte-path/some.svg.d.ts
    function fileExists(fileName) {
        const transformer = findConfig(fileName);
        return !!transformer || compilerHost.fileExists(fileName);
    }
    // When fileExists() returns true
    // CompilerHost request the source file by getSourceFile()
    function getSourceFile(fileName, languageVersion, onError, shouldCreateNewSourceFile) {
        const transformer = findConfig(fileName);
        // if transformConfig exists this CompilerHost returns the transformed data
        if (transformer) {
            const option = /\.svg\.tsx$/.test(fileName)
                ? {
                    variant: process.env.TS_SVG_EXPORT === 'default'
                        ? 'default'
                        : 'create-react-app',
                }
                : undefined;
            const sourceText = transformer.getSourceText(option)(
            // remove .tsx
            fileName.substr(0, fileName.length - 4));
            return typescript_1.default.createSourceFile(fileName, sourceText, options.target || typescript_1.default.ScriptTarget.Latest, setParentNodes, typescript_1.default.ScriptKind.TSX);
        }
        // if not, pass to the original CompilerHost
        return compilerHost.getSourceFile(fileName, languageVersion, onError, shouldCreateNewSourceFile);
    }
    return {
        ...compilerHost,
        fileExists,
        getSourceFile,
    };
}
exports.createExtendedCompilerHost = createExtendedCompilerHost;
