// @ssen/collect-dependencies/collectDependencies.js
import { builtinModules } from 'module';
import ts from 'typescript';
const nodeAPIList = new Set(builtinModules);
export const collectTypeScript = {
    extensions: ['.ts', '.tsx'],
    excludes: [
        // exclude tests
        '**/*.(spec|test).(js|jsx|ts|tsx)',
        '**/__*',
        // exclude public
        '**/public',
        '**/bin',
        // exclude javascript
        '**/*.js',
        '**/*.jsx',
    ],
    includes: ['**/*'],
};
export const collectScripts = {
    extensions: ['.js', '.jsx', '.ts', '.tsx'],
    excludes: [
        // exclude tests
        '**/*.(spec|test).(js|jsx|ts|tsx)',
        '**/__*',
        // exclude public
        '**/public',
        '**/bin',
        // exclude public
        '**/public',
    ],
    includes: ['**/*'],
};
export async function collectDependencies({ rootDir, internalPackages = new Map(), externalPackages, extensions = collectTypeScript.extensions, excludes = collectTypeScript.excludes, includes = collectTypeScript.includes, compilerOptions = {}, selfNames = new Set(), fixImportPath, checkUndefinedPackage = 'error', }) {
    compilerOptions = {
        allowJs: extensions.some((ext) => /^.js/.test(ext)),
        ...compilerOptions,
        rootDir,
    };
    const host = ts.createCompilerHost(compilerOptions);
    const files = host.readDirectory(rootDir, extensions, excludes, includes);
    const program = ts.createProgram(files, compilerOptions, host);
    const importPaths = new Set();
    for (const file of files) {
        const sourceFile = program.getSourceFile(file);
        if (!sourceFile)
            continue;
        function search(node) {
            // import from '?'
            if (ts.isImportDeclaration(node) &&
                ts.isStringLiteralLike(node.moduleSpecifier) &&
                node.moduleSpecifier.text) {
                const importPath = node.moduleSpecifier.text;
                importPaths.add(typeof fixImportPath === 'function'
                    ? fixImportPath({
                        importPath,
                        filePath: file,
                    })
                    : importPath);
            }
            // import('?')
            else if (ts.isCallExpression(node) &&
                node.expression.kind === ts.SyntaxKind.ImportKeyword &&
                ts.isStringLiteralLike(node.arguments[0])) {
                const importPath = node.arguments[0]
                    .text;
                importPaths.add(typeof fixImportPath === 'function'
                    ? fixImportPath({
                        importPath,
                        filePath: file,
                    })
                    : importPath);
            }
            // require.resolve('?')
            else if (ts.isCallExpression(node) &&
                ts.isPropertyAccessExpression(node.expression) &&
                ts.isIdentifier(node.expression.expression) &&
                node.expression.expression.escapedText === 'require' &&
                node.expression.name.escapedText === 'resolve' &&
                ts.isStringLiteralLike(node.arguments[0])) {
                const importPath = node.arguments[0]
                    .text;
                importPaths.add(typeof fixImportPath === 'function'
                    ? fixImportPath({
                        importPath,
                        filePath: file,
                    })
                    : importPath);
            }
            // require('?')
            else if (ts.isCallExpression(node) &&
                ts.isIdentifier(node.expression) &&
                node.expression.escapedText === 'require' &&
                ts.isStringLiteralLike(node.arguments[0])) {
                const importPath = node.arguments[0]
                    .text;
                importPaths.add(typeof fixImportPath === 'function'
                    ? fixImportPath({
                        importPath,
                        filePath: file,
                    })
                    : importPath);
            }
            ts.forEachChild(node, search);
        }
        search(sourceFile);
    }
    const imports = {};
    for (const importPath of importPaths) {
        const packageName = /^@/.test(importPath)
            ? importPath.split('/').slice(0, 2).join('/')
            : importPath.split('/')[0];
        if (!imports[packageName] &&
            !selfNames.has(packageName) &&
            !/^\./.test(packageName) &&
            !nodeAPIList.has(packageName)) {
            const internalPackage = internalPackages.get(packageName);
            if (internalPackage) {
                imports[packageName] = `^${internalPackage.version}`;
            }
            else if (externalPackages[packageName]) {
                imports[packageName] = externalPackages[packageName];
            }
            else if (checkUndefinedPackage === 'error') {
                throw new Error(`Undefined package "${packageName}" from "${importPath}"`);
            }
        }
    }
    return imports;
}
