// rocket-punch/entry/readPackages.js
import fs from 'fs-extra';
import path from 'path';
export async function readPackages({ cwd, sourceRoot, entry, }) {
    const packages = {};
    for (const name of Object.keys(entry)) {
        if (/\/\*$/.test(name)) {
            const groupName = name.split('/')[0];
            const dir = path.resolve(cwd, sourceRoot, groupName);
            const files = await fs.readdir(dir);
            for (const pkgName of files) {
                const pkgDir = path.join(dir, pkgName);
                const groupAndPkgName = groupName + '/' + pkgName;
                if (fs.statSync(pkgDir).isDirectory() &&
                    fs.readdirSync(pkgDir).length > 0 &&
                    !packages[groupAndPkgName]) {
                    packages[groupName + '/' + pkgName] = entry[name];
                }
            }
        }
        else if (!packages[name]) {
            packages[name] = entry[name];
        }
    }
    return Object.keys(packages).reduce((map, name) => {
        var _a, _b, _c, _d;
        const versionOrInfo = packages[name];
        if (typeof versionOrInfo === 'string') {
            map.set(name, {
                name,
                version: versionOrInfo,
                tag: 'latest',
                //module: 'esm',
                exports: {
                    main: 'module',
                    module: true,
                    commonjs: true,
                },
                access: undefined,
                registry: undefined,
                compilerOptions: {},
                packageJson: {},
            });
        }
        else {
            const exports = (_a = versionOrInfo.exports) !== null && _a !== void 0 ? _a : ['module', 'commonjs'];
            const commonjsExists = Array.from(exports).includes('commonjs');
            const moduleExists = Array.from(exports).includes('module');
            if (!commonjsExists && !moduleExists) {
                throw new Error(`there are no exports! ${versionOrInfo.exports}`);
            }
            const main = exports[0];
            if (!main) {
                throw new Error(`can't resolve main`);
            }
            map.set(name, {
                name,
                version: versionOrInfo.version,
                tag: (_b = versionOrInfo.tag) !== null && _b !== void 0 ? _b : 'latest',
                //module: versionOrInfo.module === 'esm' ? 'esm' : 'commonjs',
                exports: {
                    main,
                    module: moduleExists,
                    commonjs: commonjsExists,
                },
                access: versionOrInfo.access,
                registry: versionOrInfo.registry,
                compilerOptions: (_c = versionOrInfo.compilerOptions) !== null && _c !== void 0 ? _c : {},
                packageJson: (_d = versionOrInfo.packageJson) !== null && _d !== void 0 ? _d : {},
            });
        }
        return map;
    }, new Map());
}
